### PR TITLE
Firelock lights on hot / pressurized rooms (#15852)

### PR DESCRIPTION
**Pull request by tom-leys**
## About the PR
Firelock doors check for atmos / heat differences and show closed lights when they are present (spaced rooms or fire on the other side for instance)

Detail:
- Update closed lights (DoorVisualLayers.BaseUnlit) on FireLocks
- Update firelock closed light signal when door is open (unset it)
- Enable the closed sprite layer on firelocks

**Media**
No new media has been added to the game, but existing close lights for FireLocks are now enabled (sometimes) 

image
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- fix: Fixed firelocks not lighting up when blocking fire, space, or pressure.

### CODE REVIEW

**Comment by VasilisThePikachu** [comment]
Closes #15852

File: Content.Client/Doors/DoorSystem.cs [hunk 1]
```diff
@@ -131,6 +131,18 @@ private void OnAppearanceChange(EntityUid uid, DoorComponent comp, ref Appearanc
             default:
                 throw new ArgumentOutOfRangeException($"Invalid door visual state {state}");
         }
+
+        // if we have https://github.com/space-wizards/RobustToolbox/pull/3984
+        // if (args.Sprite.LayerExists(DoorVisualLayers.BaseUnlit))
+        if (args.Sprite.LayerMapTryGet(DoorVisualLayers.BaseUnlit, out _, false))
```
Comment (ElectroJr) [inline Content.Client/Doors/DoorSystem.cs:137]: In general just use `LayerMapTryGet()` and pass the layer index to `LayerSetWhatever()`

File: Resources/Prototypes/Entities/Structures/Doors/Firelocks/firelock.yml [hunk 2]
```diff
@@ -37,7 +37,6 @@
         - state: closed_unlit
           shader: unshaded
           map: ["enum.DoorVisualLayers.BaseUnlit"]
-          visible: false
```
Comment (ElectroJr) [inline Resources/Prototypes/Entities/Structures/Doors/Firelocks/firelock.yml:40]: By default, this shouldn't be visible and you shouldn't have to remove this line.

**Review by ElectroJr — DISMISSED** [review]
The visualizer new appearance logic shouldn't be in the door system as its just for firelocks, and it might also cause issues with the Airlock appearance logic (which are all also doors).

Firelocks should probably get their own logic to check for power & toggle the layer

**Comment by tom-leys** [comment]
@ElectroJr 
> Firelocks should probably get their own logic to check for power & toggle the layer

This is currently beyond my ability. 

It seems that there is a DoorComponent and an AirlockComponent in the shared code. Then there are relevant systems in the client to work on those.

Do I need to add a shared FirelockComponent and then a new FirelockSystem on the client? 

How do I ensure that the new shared FireLock component is created by the server and then created on the client?

Thanks

-Tom

**Comment by ElectroJr** [comment]
> Do I need to add a shared FirelockComponent and then a new FirelockSystem on the client?
> 
> How do I ensure that the new shared FireLock component is created by the server and then created on the client?

I would just move the existing FirelockComponent from `Content.Server` to `Content.Shared`, and then yes you'd just have to create a client-side system called something like `FirelockSystem` or `FirelockVisualsSystem`.

Moving the component will mean the entities will already have the right components automatically. You could also have had a separate content & client component with the same name and that would also have worked, but its nicer to just have a single shared component.

**Comment by tom-leys** [comment]
@ElectroJr this is now resolved, please review

File: Content.Client/Doors/FirelockSystem.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<FirelockComponent, AppearanceChangeEvent>(OnAppearanceChange);
+    }
+
+    private void OnAppearanceChange(EntityUid uid, FirelockComponent comp, ref AppearanceChangeEvent args)
+    {
+        if (args.Sprite == null || !_gameTiming.IsFirstTimePredicted)
```
Comment (ElectroJr) [inline Content.Client/Doors/FirelockSystem.cs:21]: Appearance updates only happen during frame-updates, so there is no need to check `IsFirstTimePredicted`.

**Pull request merged**